src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,223 @@
+            );
+        }
+        this.max_window_size = max_window_size;
+    }
+
+    /**
+     * Process the search request of neural_sparse_two_phase_processor
+     * @param request the search request (which may have been modified by an earlier processor)
+     * @return request the search request that add the two-phase rescore query of neural sparse query.
+     */
+    @Override
+    public SearchRequest processRequest(final SearchRequest request) throws Exception {
+        if (!enabled || ratio == 0f) {
+            return request;
+        }
+        QueryBuilder queryBuilder = request.source().query();
+        // Collect the nested NeuralSparseQueryBuilder in the whole query.
+        Multimap<NeuralSparseQueryBuilder, Float> queryBuilderMap = ArrayListMultimap.create();
+        queryBuilderMap = collectNeuralSparseQueryBuilder(queryBuilder, 1.0f);
+        if (queryBuilderMap.isEmpty()) return request;
+        // Make a nestedQueryBuilder which includes all the two-phase QueryBuilder.
+        QueryBuilder nestedTwoPhaseQueryBuilder = getNestedQueryBuilderFromNeuralSparseQueryBuilderMap(queryBuilderMap);
+        nestedTwoPhaseQueryBuilder.boost(getOriginQueryWeightAfterRescore(request.source()));
+        // Add it to the rescorer.
+        RescorerBuilder<QueryRescorerBuilder> twoPhaseRescorer = buildRescoreQueryBuilderForTwoPhase(nestedTwoPhaseQueryBuilder,request);
+        request.source().addRescorer(twoPhaseRescorer);
+        return request;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    /**
+     * Based on ratio, split a Map into two map by the value.
+     * @param queryTokens the queryTokens map, key is the token String, value is the score.
+     * @param ratio The ratio that control how tokens map be split.
+     * @return A map has two element, {[True, token map whose value above threshold],[False, token map whose value below threshold]}
+     */
+    public static Map<Boolean, SetOnce<Map<String, Float>>> getSplitSetOnceByScoreThreshold(
+        final Map<String, Float> queryTokens,
+        final float ratio
+    ) {
+        float max = 0f;
+        for (Float value : queryTokens.values())
+            max = Math.max(value, max);
+        float threshold = max * ratio;
+
+        Map<Boolean, Map<String, Float>> queryTokensByScore = queryTokens.entrySet()
+            .stream()
+            .collect(
+                Collectors.partitioningBy(entry -> entry.getValue() >= threshold, Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+            );
+        SetOnce<Map<String, Float>> highScoreTksSetOnce = new SetOnce<>(queryTokensByScore.get(Boolean.TRUE));
+        SetOnce<Map<String, Float>> lowScoreTksSetOnce = new SetOnce<>(queryTokensByScore.get(Boolean.FALSE));
+        if (highScoreTksSetOnce.get() == null) throw new IllegalArgumentException();
```

Comment:
```suggestion
        if (Objects.isNull(highScoreTksSetOnce.get())) {
                 throw new IllegalArgumentException();
        }         
```

Reply:
In which case the highScoreTks will be null? And why it's an IllegalArgumentException, also add error message to exceptions.

Reply:
I think only when the model inference error or the parameter is null or empty, the highScoreTks will be null.
And a simple error msg now.